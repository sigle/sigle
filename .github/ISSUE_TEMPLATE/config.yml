blank_issues_enabled: false
contact_links:
  - name: ðŸ™‹ Ask a question
    url: https://app.sigle.io/discord
    about: Ask questions and discuss with other community members
