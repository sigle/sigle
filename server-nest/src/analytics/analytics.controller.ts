import { Controller } from '@nestjs/common';

@Controller('analytics')
export class AnalyticsController {}
