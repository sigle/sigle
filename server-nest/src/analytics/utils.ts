import { isBefore } from 'date-fns';
import { fetch } from 'undici';
import { migrationStories, SubsetStory } from '../external/gaia';

// All queries should start at this date maximum.
const STATS_MAX_FROM_DATE = '2022-05-01';

/**
 * Set max date in the past. Fathom data is not correct before this one.
 */
export const maxFathomFromDate = (parsedDateFrom: Date, dateFrom: string) => {
  if (isBefore(parsedDateFrom, new Date(STATS_MAX_FROM_DATE))) {
    return STATS_MAX_FROM_DATE;
  }
  return dateFrom;
};

export const getPublicStories = async ({
  bucketUrl,
}: {
  bucketUrl: string;
}): Promise<SubsetStory[]> => {
  const resPublicStories = await fetch(`${bucketUrl}publicStories.json`);
  // This would happen if the user has not published any stories
  if (resPublicStories.status !== 200) {
    return [];
  }
  const publicStoriesFile = migrationStories(
    (await resPublicStories.json()) as any,
  );
  return publicStoriesFile.stories;
};
