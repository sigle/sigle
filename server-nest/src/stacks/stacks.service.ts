import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';
import { fetch } from 'undici';

@Injectable()
export class StacksService {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}
  /**
   * Return the username of a stacks address.
   * Cache the result for better performance as the username is not changing often.
   */
  async getUsernameByAddress(address: string): Promise<string> {
    const cacheKey = `stacks_username:${address}`;
    const cachedResponse = await this.cacheManager.get<string>(cacheKey);
    if (cachedResponse) {
      return cachedResponse;
    }

    let username: string;
    const namesResponse = await fetch(
      `https://stacks-node-api.stacks.co/v1/addresses/stacks/${address}`,
    );
    const namesJson = (await namesResponse.json()) as { names: string[] };
    if ((namesJson.names.length || 0) > 0) {
      username = namesJson.names[0];
    } else {
      throw new Error(`No username found for ${address}`);
    }

    // Cache response for 1 day
    await this.cacheManager.set(cacheKey, username, 24 * 60 * 60);
    return username;
  }
}
