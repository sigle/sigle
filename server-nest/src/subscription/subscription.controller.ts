import { Controller, Get, Request, UseGuards } from '@nestjs/common';
import { ApiBearerAuth, ApiOkResponse, ApiOperation } from '@nestjs/swagger';
import { AuthGuard } from '../auth.guard';
import { SubscriptionDto } from './dto/subscription.dto';
import { SubscriptionService } from './subscription.service';

@Controller()
export class SubscriptionController {
  constructor(private readonly subscriptionService: SubscriptionService) {}

  @ApiOperation({
    description:
      'Return the current active subscription of the current logged in user.',
  })
  @ApiBearerAuth()
  @ApiOkResponse({
    description:
      'Returns the current active subscription object. If no active subscription is found, null is returned.',
    type: SubscriptionDto,
  })
  @UseGuards(AuthGuard)
  @Get('/api/subscriptions')
  getUserMe(@Request() req): Promise<SubscriptionDto> {
    return this.subscriptionService.getUserActiveSubscription({
      stacksAddress: req.address,
    });
  }
}
