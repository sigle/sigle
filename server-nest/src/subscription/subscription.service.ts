import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma.service';

@Injectable()
export class SubscriptionService {
  constructor(private prisma: PrismaService) {}

  async getUserActiveSubscription({
    stacksAddress,
  }: {
    stacksAddress: string;
  }) {
    const activeSubscription = await this.prisma.subscription.findFirst({
      where: {
        user: { stacksAddress },
        status: 'ACTIVE',
      },
      select: {
        id: true,
        nftId: true,
      },
    });
    return activeSubscription;
  }
}
