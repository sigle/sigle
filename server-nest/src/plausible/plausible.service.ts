import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { EnvironmentVariables } from '../environment/environment.validation';

@Injectable()
export class PlausibleService {
  private siteId: string;
  private apiToken: string;

  constructor(
    private readonly configService: ConfigService<EnvironmentVariables>,
  ) {
    this.siteId = this.configService.get('PLAUSIBLE_SITE_ID');
    this.apiToken = this.configService.get('PLAUSIBLE_API_TOKEN');
  }

  async pages(params: {
    dateFrom: string;
    dateTo: string;
    paths: string[];
  }): Promise<
    {
      page: string;
      pageviews: number;
      visitors: number;
    }[]
  > {
    const urlParams = new URLSearchParams({
      period: 'custom',
      date: `${params.dateFrom},${params.dateTo}`,
      property: 'event:page',
      filters: `event:page==${params.paths.join('|')}`,
      metrics: 'visitors,pageviews',
    });

    const data = await this.fetch('/stats/breakdown', urlParams);
    return data.results;
  }

  async referrers(params: {
    dateFrom: string;
    dateTo: string;
    paths: string[];
  }): Promise<
    {
      source: string;
      pageviews: number;
      visitors: number;
    }[]
  > {
    const urlParams = new URLSearchParams({
      period: 'custom',
      date: `${params.dateFrom},${params.dateTo}`,
      property: 'visit:source',
      filters: `event:page==${params.paths.join('|')}`,
      metrics: 'visitors,pageviews',
    });

    const data = await this.fetch('/stats/breakdown', urlParams);
    return data.results;
  }

  async timeseries(params: {
    dateGrouping: 'day' | 'month';
    dateFrom: string;
    dateTo: string;
    paths: string[];
  }): Promise<
    {
      date: string;
      pageviews: number;
      visitors: number;
    }[]
  > {
    const urlParams = new URLSearchParams({
      period: 'custom',
      interval: params.dateGrouping === 'day' ? 'date' : 'month',
      date: `${params.dateFrom},${params.dateTo}`,
      filters: `event:page==${params.paths.join('|')}`,
      metrics: 'visitors,pageviews',
    });

    const data = await this.fetch('/stats/timeseries', urlParams);
    return data.results;
  }

  private async fetch(path: string, params: URLSearchParams): Promise<any> {
    params.append('site_id', this.siteId);

    const response = await fetch(
      `https://plausible.io/api/v1${path}?${params.toString()}`,
      {
        headers: {
          Authorization: `Bearer ${this.apiToken}`,
        },
      },
    );

    if (response.status !== 200) {
      console.error(await response.text());
      throw new Error(`Failed to fetch plausible api: ${response.status}`);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return response.json() as any;
  }
}
