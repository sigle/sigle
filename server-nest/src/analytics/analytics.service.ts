import {
  BadRequestException,
  CACHE_MANAGER,
  Inject,
  Injectable,
} from '@nestjs/common';
import { Cache } from 'cache-manager';
import { format, isValid, parse } from 'date-fns';
import { PlausibleService } from 'src/plausible/plausible.service';
import { StacksService } from '../stacks/stacks.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { getPublicStories, maxFathomFromDate } from './utils';

@Injectable()
export class AnalyticsService {
  constructor(
    @Inject(CACHE_MANAGER) private readonly cacheManager: Cache,
    private readonly subscriptionService: SubscriptionService,
    private readonly stacksService: StacksService,
    private readonly plausibleService: PlausibleService,
  ) {}

  async referrers({
    stacksAddress,
    dateFrom,
    storyId,
  }: {
    stacksAddress: string;
    dateFrom: string;
    storyId?: string;
  }) {
    const dateTo = new Date();

    const parsedDateFrom = parse(dateFrom, 'yyyy-MM-dd', new Date());
    const isValidDate = isValid(parsedDateFrom);
    if (!isValidDate) {
      throw new BadRequestException('Invalid date from.');
    }

    // Set max date in the past
    dateFrom = maxFathomFromDate(parsedDateFrom, dateFrom);

    const activeSubscription =
      await this.subscriptionService.getUserActiveSubscription({
        stacksAddress,
      });
    if (!activeSubscription) {
      throw new BadRequestException('No active subscription.');
    }

    const username = await this.stacksService.getUsernameByAddress(
      stacksAddress,
    );

    // Caching mechanism to avoid hitting Plausible too often
    const cacheKey = storyId
      ? `referrers:${username}_${dateFrom}_${storyId}`
      : `referrers:${username}_${dateFrom}`;
    const cachedResponse = await this.cacheManager.get<string>(cacheKey);
    if (cachedResponse) {
      return JSON.parse(cachedResponse);
    }

    const { profile, bucketUrl } = await this.stacksService.getBucketUrl({
      username,
    });
    if (!profile || !bucketUrl) {
      throw new BadRequestException(`No profile or bucketUrl for ${username}.`);
    }

    let storiesPath: string[];
    if (storyId) {
      storiesPath = [`/${username}/${storyId}`];
    } else {
      const publicStoriesFile = await getPublicStories({ bucketUrl });
      storiesPath = publicStoriesFile.map(
        (publicStory) => `/${username}/${publicStory.id}`,
      );
      // Add the root path to the list of paths
      storiesPath.push(`/${username}`);
    }

    const plausibleReferrers = await this.plausibleService.referrers({
      dateFrom,
      dateTo: format(dateTo, 'yyyy-MM-dd'),
      paths: storiesPath,
    });

    const referrersResponse: {
      domain: string | null;
      count: number;
    }[] = plausibleReferrers
      .map((result) => ({
        domain: result.source,
        count: result.pageviews,
      }))
      .sort((a, b) => b.count - a.count);

    // Cache response for 1 hour
    await this.cacheManager.set(
      cacheKey,
      JSON.stringify(referrersResponse),
      60 * 60,
    );

    return referrersResponse;
  }
}
