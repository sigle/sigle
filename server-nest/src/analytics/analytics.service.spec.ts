import { CACHE_MANAGER } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { PlausibleService } from '../plausible/plausible.service';
import { StacksService } from '../stacks/stacks.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { AnalyticsService } from './analytics.service';

describe('AnalyticsService', () => {
  let service: AnalyticsService;
  let plausibleService: PlausibleService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AnalyticsService,
        {
          provide: CACHE_MANAGER,
          useValue: {
            get: () => null,
            set: () => () => null,
          },
        },
        {
          provide: ConfigService,
          useValue: {
            get: jest.fn((key: string) => {
              return null;
            }),
          },
        },
        {
          provide: SubscriptionService,
          useValue: {
            getUserActiveSubscription: () =>
              jest.fn(() => ({
                id: 'test',
              })),
          },
        },
        {
          provide: StacksService,
          useValue: {
            getUsernameByAddress: () => 'sigleapp.id.blockstack',
            getBucketUrl: () => ({
              profile: {},
              bucketUrl: 'https://www.sigle.io',
            }),
            getPublicStories: () => [],
          },
        },
        {
          provide: PlausibleService,
          useValue: {
            referrers: jest.fn(),
          },
        },
      ],
    }).compile();

    service = module.get<AnalyticsService>(AnalyticsService);
    plausibleService = await module.resolve(PlausibleService);
  });

  describe('referrers', () => {
    it('should return referrers', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.referrers as jest.Mock).mockResolvedValueOnce([
        {
          pageviews: '1',
          source: 'https://stacks.co',
        },
        {
          pageviews: '10',
          source: 'Direct / None',
        },
      ]);

      const data = await service.referrers({
        stacksAddress,
        dateFrom: '2022-03-15',
      });
      expect(plausibleService.referrers).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(data).toMatchSnapshot();
    });

    it('should return referrers for one story', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.referrers as jest.Mock).mockResolvedValueOnce([
        {
          pageviews: '5',
          source: 'https://stacks.co',
        },
      ]);

      const data = await service.referrers({
        stacksAddress,
        dateFrom: '2022-03-15',
        storyId: 'test',
      });
      expect(plausibleService.referrers).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: ['/sigleapp.id.blockstack/test'],
      });
      expect(data).toMatchSnapshot();
    });
  });
});
