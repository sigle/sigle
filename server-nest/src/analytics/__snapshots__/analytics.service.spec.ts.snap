// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyticsService referrers should return referrers 1`] = `
[
  {
    "count": "10",
    "domain": "Direct / None",
  },
  {
    "count": "1",
    "domain": "https://stacks.co",
  },
]
`;

exports[`AnalyticsService referrers should return referrers for one story 1`] = `
[
  {
    "count": "5",
    "domain": "https://stacks.co",
  },
]
`;
